common/database/src/main/kotlin/no/nav/mulighetsrommet/database/DatabaseAdapter.kt

```diff
@@ -83,39 +81,63 @@ open class DatabaseAdapter(config: DatabaseConfig) : Database {
         return createArrayOf("integer", list)
     }
 
-    override fun <T> run(query: NullableResultQueryAction<T>): T? {
-        return using(session) {
+    override fun <T> run(query: NullableResultQueryAction<T>): T? =
+        useSession {
             it.run(query)
         }
-    }
 
-    override fun <T> run(query: ListResultQueryAction<T>): List<T> {
-        return using(session) {
+    override fun <T> run(query: ListResultQueryAction<T>): List<T> =
+        useSession {
             it.run(query)
         }
-    }
 
-    override fun run(query: ExecuteQueryAction): Boolean {
-        return using(session) {
+    override fun run(query: ExecuteQueryAction): Boolean =
+        useSession {
             it.run(query)
         }
-    }
 
-    override fun run(query: UpdateQueryAction): Int {
-        return using(session) {
+    override fun run(query: UpdateQueryAction): Int =
+        useSession {
             it.run(query)
         }
-    }
 
-    override fun run(query: UpdateAndReturnGeneratedKeyQueryAction): Long? {
-        return using(session) {
+    override fun run(query: UpdateAndReturnGeneratedKeyQueryAction): Long? =
+        useSession {
             it.run(query)
         }
-    }
 
     override fun <T> transaction(operation: (TransactionalSession) -> T): T {
-        return using(session) {
-            it.transaction(operation)
+        val tx = transactionalSession
+
+        return if (tx != null) {
+            operation.invoke(tx)
+        } else {
+            using(sessionOf(dataSource, strict = true)) {
+                try {
+                    it.connection.begin()
+                    val tx = TransactionalSession(it.connection, it.returnGeneratedKeys, it.autoGeneratedKeys, it.strict)
+                    transactionalSession = tx
+                    val result = operation.invoke(tx)
+                    it.connection.commit()
+                    result
+                } catch (e: Exception) {
+                    it.connection.rollback()
+                    throw e
+                } finally {
+                    transactionalSession = null
+                }
+            }
+        }
+    }
```

Comment:
Dette er basically en kopi av session.transaction metoden, men lagrer transactionSession'en i klassen for at db.run kall skal bruke samme session. Man kunne i stedet skrevet om alle db metodene til å ta inn en session 🤔 Evt. brukt et annet jdbc bibliotek 🤔